### ISSUE TITLE
Encoder loops over views one at a time, starving the GPU (5.2x slower than folding)

### ISSUE THREAD

**Issue by engichang1467** [engichang1467/Open-Rig3R#33]
## Problem

`Rig3R.forward` encoded each view in a Python loop, one `self.encoder` call per view:

```python
# models/rig3r.py (before)
tokens_list = []
for i in range(N):
    enc_out = self.encoder(images[:, i])
    tokens_list.append(enc_out["tokens"])   # (B, num_patches, C)
joint_tokens = torch.cat(tokens_list, dim=1)
```

The encoder is per-image and stateless across views, so the loop buys nothing. With the
Waymo full rig, `N = n_frames * num_cameras = 2 * 5 = 10`, so a `batch_size: 4` step
issued **10 sequential ViT forward passes of only 4 images each**. Each launch is far too
small to fill an A100's SMs, and the 10 launches serialize, so the GPU idles between them.

## Fix

Fold views into the batch dimension and encode once:

```python
# models/rig3r.py (after)
tokens = self.encoder(images.reshape(B * N, C, H, W))["tokens"]   # (B*N, num_patches, C)
joint_tokens = tokens.reshape(B, N * tokens.shape[1], tokens.shape[2])
```

The fold's row order is `(b0v0, b0v1, ..., b1v0, ...)`, so the reshape back reproduces
exactly the per-view concatenation the decoder already expects. No change to math or
to `RigAwareTransformerDecoder`.

## Measured

Nvidia GPU (ex. A100 80GB, or H100), `batch_size=4`, `n_frames=2`, 5 cameras (V=10), 128x128, bf16 autocast,
full forward + backward + optimizer step, median of 10 steps after warmup:

| path | ms/step | samples/s | peak memory |
|---|---|---|---|
| per-view loop | 822 | 4.9 | 6.1 GiB |
| folded | ~150 | ~27 | 6.1 GiB |

**5.2x faster at identical peak memory.** Memory is unchanged because the loop also had
to retain every view's activations for backward — nothing was being saved by going one
view at a time.

Note: an earlier informal measurement of this reported 141 ms / 5.8x. Benchmarking both
paths in the same process showed whichever ran *second* gained ~50 ms from allocator and
cuDNN warmup; swapping the order swapped which one looked faster. The apples-to-apples
figure is ~150 ms / 5.2x.

## Portability

Not A100-specific — it is a kernel-launch/occupancy win on any NVIDIA GPU. The
speedup shrinks on smaller cards that a `B*N` batch already saturates (an RTX 5070 has
roughly half the SMs, so expect ~2-3x), but it is never a regression, and peak memory is
identical either way.

## Test

`tests/test_rig3r_forward.py::test_view_fold_matches_per_view_loop` reconstructs the old
loop-and-concat path from the same weights and asserts every output head matches the
folded path via `torch.testing.assert_close`. Passes.
